lightningnetwork/lnd
# Specify macaroon root key during wallet init

## Change Description

This set of commits adds the ability to specify the macaroon root key to be used when initializing a wallet. This allows for deterministically generating macaroons for a new node/wallet. In tandem with specifying the wallet seed, a wallet can be initialized without requiring secrets to be randomly generated by lnd upon wallet initialization.

## Steps to Test

These are the steps I used to manually test this change.

1. Start lnd from a fresh start with no wallet.
2. Use the commands below to initialize the wallet with a root macaroon key we randomly generate outside of lnd.

```bash
CA_CERT=~/.lnd/tls.cert

WALLETPASS=somepass
WALLETPASS_BASE64=$(printf $WALLETPASS|base64|tr -d '\n\r')

ROOTKEY_BASE64=$(dd status=none if=/dev/urandom bs=32 count=1 | base64)
echo $ROOTKEY_BASE64

CIPHER_SEED=$(curl -s --cacert "$CA_CERT" -X GET https://localhost:8080/v1/genseed | jq -c -r '.cipher_seed_mnemonic')

INITWALLET_REQ='{"wallet_password":"'$WALLETPASS_BASE64'", "macaroon_root_key":"'$ROOTKEY_BASE64'", "cipher_seed_mnemonic":'$CIPHER_SEED'}'

curl -s --cacert "$CA_CERT" -X POST -d "$INITWALLET_REQ" https://localhost:8080/v1/initwallet > /dev/null
```
3. Verify that you can use the newly generated `admin.macaroon` to make `lncli` commands to lnd.
4. Rename `admin.macaroon` to `admin.macaroon.old` and move it outside of the lnd dir.
5. Delete the lnd `data` dir
6. Start lnd again from another fresh start.
7. Run the same commands as in step 2 to init the wallet but assign `ROOTKEY_BASE64` to the value that was `echo`ed in step 2, e.g. `ROOTKEY_BASE64="3BVV9mZmqAv0/BUS2jOvKLNDGhxssLTAfVlUYGeSAUw="`
8. Verify that you can make `lncli` commands with the `admin.macaroon.old` file you saved from step 4.
